### PR TITLE
Refine test case Prepare and Cleanup methods

### PR DESCRIPTION
**Pull request by scothis**
This is a breaking API change, existing uses of the Prepare and Cleanup
methods were half baked and uncommon. While the API signature is
changing the existing capabilities are preserved.

The Prepare method signature now takes the form:

```
func (t *testing.T, ctx context.Context, tc *XTestCase) (context.Context, error)
```

The provided context may be augmented by the Prepare method, and then
must be returned. The XTestCase is one of `ReconcilerTestCase`,
`SubReconcilerTestCase` or `AdmissionWebhookTestCase`. The method is
called before any other state for the test has been allocated, this
means the test case may be manipulated.

The Cleanup method signature now takes the form:

```
func (t *testing.T, ctx context.Context, tc *XTestCase) error
```

Values set on the context in the prepare method are available here, as
well as any values stashed in the test case metadata.

This change also consistently set a test logr into the context. This
will show reconciler logs for test failures. This behavior was already
in place for AdmissionWebhookTestCases. The context also now has a
deadline set if the test is executed with a timeout.

Signed-off-by: Scott Andrews <andrewssc@vmware.com>

### CODE REVIEW

**Comment by scothis** [comment]
cc @emmjohnson @squeedee I can't tag you for reviews since you're not a member or the org

**Comment by squeedee** [comment]
We'll try writing a test that includes log assertions and see how that goes.

File: testing/reconciler.go [hunk 1]
```diff
@@ -119,6 +122,30 @@ func (tc *ReconcilerTestCase) Run(t *testing.T, scheme *runtime.Scheme, factory
 	}
 
 	ctx := context.Background()
+	ctx = logr.NewContext(ctx, logrtesting.NewTestLogger(t))
+	if deadline, ok := t.Deadline(); ok {
```
Comment (mamachanko) [inline testing/reconciler.go:126]: **question:** Should the go doc for each TestCase mention that the ctx is cancelable?

Comment (scothis) [reply testing/reconciler.go:126]: Probably not. I'm not sure how to communicate the meaning short of saying "it's a context, treat it is as such".

It's not user cancelable, and all contexts passed into a method are inherently potentially cancelable. There's a lot of discussion upstream in https://github.com/golang/go/issues/28135. The main benefit is if there are any go routines still in flight after the test returns they can react to the context being canceled (we don't use parallel tests). Previously we used a Background context that would remain open forever, until the test runner stops the execution dead.

Comment (mamachanko) [reply testing/reconciler.go:126]: >Probably not. I'm not sure how to communicate the meaning short of saying "it's a context, treat it is as such".

`+1`

**Review by squeedee — APPROVED** [review]
Working well for me. Using this to mock logs and other constructional objects.

**Pull request merged**